cordapp-cpk/src/test/kotlin/net/corda/plugins/cpb/CpbTest.kt

```diff
@@ -0,0 +1,139 @@
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.BeforeAll
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.TestReporter
+import org.junit.jupiter.api.io.TempDir
+import java.io.IOException
+import java.io.InputStream
+import java.nio.file.*
+import java.nio.file.attribute.BasicFileAttributes
+import java.security.DigestInputStream
+import java.security.MessageDigest
+import java.security.cert.X509Certificate
+import java.util.*
+import java.util.jar.JarInputStream
+import javax.security.auth.x500.X500Principal
+
+class CpbTest {
+    companion object {
+        private const val cordappVersion = "1.2.1"
+        private const val hostVersion = "2.0.1"
+        private val cordaDevCertPrincipal = X500Principal("CN=Corda Dev Code Signer, OU=R3, O=Corda, L=London, C=GB")
+
+
+        private lateinit var externalProject: GradleProject
+        private lateinit var testProject: GradleProject
+        private lateinit var testDir: Path
+        private lateinit var mavenRepoDir : Path
+        private lateinit var externalCordappProjectDir : Path
+        private lateinit var cpbProjectDir : Path
+
+        @Suppress("unused")
+        @BeforeAll
+        @JvmStatic
+        fun setup(@TempDir tmpDir: Path, reporter: TestReporter) {
+            testDir = tmpDir
+            mavenRepoDir = Files.createDirectory(testDir.resolve("maven"))
+            externalCordappProjectDir = Files.createDirectory(testDir.resolve("externalCordapp"))
+            cpbProjectDir = Files.createDirectory(testDir.resolve("cpb"))
+            externalProject = GradleProject(externalCordappProjectDir, reporter)
+                .withTestName("external-cordapp")
+                .withTaskName("publishAllPublicationsToTestRepository")
+                .build("-Pmaven.repository.dir=$mavenRepoDir",
+                    "-Phost_version=$hostVersion",
+                    "-Pcordapp_version=$cordappVersion",
+                    "-Pcordapp_contract_version=$expectedCordappContractVersion")
+            testProject = GradleProject(cpbProjectDir, reporter)
+                .withTestName("cordapp-cpb")
+                .withSubResource("project-dependency/build.gradle")
+                .withTaskName("cpb")
+                .build("-Pmaven.repository.dir=$mavenRepoDir",
+                    "-Pcordapp_version=$cordappVersion",
+                    "-Phost_version=$hostVersion",
+                    "-Pcordapp_contract_version=$expectedCordappContractVersion")
+        }
+
+        private val HEX_ARRAY = "0123456789ABCDEF".toCharArray()
+        private fun ByteArray.toHex(): String {
+            val hexChars = CharArray(size * 2)
+            for (j in indices) {
+                val v: Int = this[j].toInt() and 0xFF
+                hexChars[j * 2] = HEX_ARRAY[v ushr 4]
+                hexChars[j * 2 + 1] = HEX_ARRAY[v and 0x0F]
+            }
+            return String(hexChars)
+        }
```

Comment:
this is a lot slower, almost as slow as copying all the test resource files in the junit temporary directory XP

Reply:
I haven't benchmarked either, but this version is considerably _shorter_. And if Kotlin provided a built-in `ByteArray.toHex()` function then I would ask you to use that, regardless of how it were implemented.

By the same metric, not writing a lot of unnecessary code to copy resource files around is "shorter" too.

Reply:
I don't need to benchmark it to tell that allocating a new string for every byte (and parsing the format string) is waaay slower than the above code, it also absolutely irrelevant given it will be executed a dozen of times at most on 32 bytes arrays, I'll amend it (it's a shame this is not in kotlin stdlib)

Reply:
I'll probably just use Base64 in the future, which is part of Java stdlib (and is also more compact)

Reply:
> I'll probably just use Base64 in the future, which is part of Java stdlib (and is also more compact)

Base64 is fine too. As you point out, any difference in performance between base64 and hexadecimal is irrelevant because all that matters here is that you are creating two `Map<String, String>` for `AbstractMap.equals()` to compare.